From 88ce18cf11d90f7c8b9afb76c03c250b977e1d05 Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Mon, 5 Jun 2023 18:37:14 +0800
Subject: [PATCH 247/263] [Huawei] Revert "[Backport] x86/microcode/AMD: Fix
 mixed steppings support"

Offering: HULK
hulk inclusion
category: others
bugzilla: 188839

--------------------------------

In the commit <d9ec9510645a>, the amd_ucode_patch is expanded as
two-dimensional array. When CONFIG_MAXSMP is enabled or CONFIG_NODES_SHIFT
is set as 10, amd_ucode_patch array would occupy memory up to 12M.

So make decision to roll back this commit.

This reverts commit d9ec9510645ab8488e033afc56d283868e860a01.

Signed-off-by: Zeng Heng <zengheng4@huawei.com>
---
 arch/x86/kernel/cpu/microcode/amd.c | 34 +++++++++++------------------
 1 file changed, 13 insertions(+), 21 deletions(-)

diff --git a/arch/x86/kernel/cpu/microcode/amd.c b/arch/x86/kernel/cpu/microcode/amd.c
index d3bce6d380ed..95b398df30bd 100644
--- a/arch/x86/kernel/cpu/microcode/amd.c
+++ b/arch/x86/kernel/cpu/microcode/amd.c
@@ -55,9 +55,7 @@ struct cont_desc {
 };
 
 static u32 ucode_new_rev;
-
-/* One blob per node. */
-static u8 amd_ucode_patch[MAX_NUMNODES][PATCH_MAX_SIZE];
+static u8 amd_ucode_patch[PATCH_MAX_SIZE];
 
 /*
  * Microcode patch container file is prepended to the initrd in cpio
@@ -431,7 +429,7 @@ apply_microcode_early_amd(u32 cpuid_1_eax, void *ucode, size_t size, bool save_p
 	patch	= (u8 (*)[PATCH_MAX_SIZE])__pa_nodebug(&amd_ucode_patch);
 #else
 	new_rev = &ucode_new_rev;
-	patch	= &amd_ucode_patch[0];
+	patch	= &amd_ucode_patch;
 #endif
 
 	desc.cpuid_1_eax = cpuid_1_eax;
@@ -577,10 +575,10 @@ int __init save_microcode_in_initrd_amd(unsigned int cpuid_1_eax)
 
 void reload_ucode_amd(unsigned int cpu)
 {
-	u32 rev, dummy __always_unused;
 	struct microcode_amd *mc;
+	u32 rev, dummy __always_unused;
 
-	mc = (struct microcode_amd *)amd_ucode_patch[cpu_to_node(cpu)];
+	mc = (struct microcode_amd *)amd_ucode_patch;
 
 	rdmsr(MSR_AMD64_PATCH_LEVEL, rev, dummy);
 
@@ -848,8 +846,6 @@ static enum ucode_state __load_microcode_amd(u8 family, const u8 *data,
 
 static enum ucode_state load_microcode_amd(u8 family, const u8 *data, size_t size)
 {
-	struct cpuinfo_x86 *c;
-	unsigned int nid, cpu;
 	struct ucode_patch *p;
 	enum ucode_state ret;
 
@@ -862,23 +858,19 @@ static enum ucode_state load_microcode_amd(u8 family, const u8 *data, size_t siz
 		return ret;
 	}
 
-	for_each_node(nid) {
-		cpu = cpumask_first(cpumask_of_node(nid));
-		c = &cpu_data(cpu);
-
-		p = find_patch(cpu);
-		if (!p)
-			continue;
-
-		if (c->microcode >= p->patch_id)
-			continue;
+	p = find_patch(0);
+	if (!p) {
+		return ret;
+	} else {
+		if (boot_cpu_data.microcode >= p->patch_id)
+			return ret;
 
 		ret = UCODE_NEW;
-
-		memset(&amd_ucode_patch[nid], 0, PATCH_MAX_SIZE);
-		memcpy(&amd_ucode_patch[nid], p->data, min_t(u32, p->size, PATCH_MAX_SIZE));
 	}
 
+	memset(amd_ucode_patch, 0, PATCH_MAX_SIZE);
+	memcpy(amd_ucode_patch, p->data, min_t(u32, p->size, PATCH_MAX_SIZE));
+
 	return ret;
 }
 
-- 
2.17.1

